velox/common/memory/MemoryUsageTracker.cpp

```diff
@@ -53,24 +69,30 @@ void MemoryUsageTracker::updateInternal(UsageType type, int64_t size) {
   // We track the peak usage of total memory independent of user and
   // system memory since freed user memory can be reallocated as system
   // memory and vice versa.
-  int64_t totalBytes = getCurrentUserBytes() + getCurrentSystemBytes();
+  int64_t totalBytes = usage(currentUsageInBytes_, UsageType::kUserMem) +
+      usage(currentUsageInBytes_, UsageType::kSystemMem);
 
   // Enforce the limit. Throw VeloxMemoryCapException exception if the limits
   // are exceeded.
   if (size > 0 &&
-      (newPeak > usage(maxMemory_, type) || totalBytes > total(maxMemory_))) {
-    // Exceeded the limit. Fail allocation after reverting changes to
-    // parent and currentUsageInBytes_.
-    if (parent_) {
-      parent_->updateInternal(type, -size);
+      (newUsage > usage(maxMemory_, type) || totalBytes > total(maxMemory_))) {
+    bool limitRaised = false;
+    if (growCallback_) {
+      limitRaised = growCallback_(type, size, *this);
+    }
+    if (!limitRaised) {
+      // Exceeded the limit.  revert the change to current usage.
+      if (parent_) {
+        parent_->updateInternal(type, -size);
```

Comment:
couldn't this trigger growCallback in the parent as well? Maybe split the functions in an updateInternal() and updateInternalUnsafe() (or similar), the ensure the latter wouldn't do the same checks and potentially trigger the callback?